src/v/storage/index_state.h

```diff
@@ -90,6 +92,9 @@ struct index_state {
 
     static std::optional<index_state> hydrate_from_buffer(iobuf);
     friend std::ostream& operator<<(std::ostream&, const index_state&);
+
+    void serde_write(iobuf&) const;
+    friend void read_nested(iobuf_parser&, index_state&, const size_t);
```

Comment:
@felixguendling maybe I need to simply understand things a little more, but I find it weird that for the low-level read nested version I can't have a member function.

Reply:
serde_write requires that it be used on a type that is an envelope (which index_state is). and read_nested works on all types and can be thought of as the most basic/general form of read_nested and its used in this case because we are dealing with instances of index_state that are deprecated so we need a lot more control.

Reply:
read_nested is necessary for lots of things, but there isn't any reason why we can't handle this specific case by having a special version where `this` is implied and the interface doesn't take `index_state&`